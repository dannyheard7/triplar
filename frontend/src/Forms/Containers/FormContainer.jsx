import React from "react";
import {inputParsers} from "Forms/utils/inputParsers.js";

export default class FormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            errors: []
        };

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        const data = this.parseFormData(event);

        this.props.apiFunction(data).then(response => {
            if (response.status === 400) {
                this.setState({errors: response.data})
            } else {
                this.setState({errors: []});
                this.props.onSuccess(response.data);
            }
        });
    }

    parseFormData(event) {
        const form = event.target;
        const data = new FormData(event.target);

        for (let name of data.keys()) {
            const input = form.elements[name];
            const parserName = input.dataset.parse;

            if (parserName) {
                const parser = inputParsers[parserName];
                const parsedValue = parser(data.get(name));
                data.set(name, parsedValue);
            }
        }

        return data;
    }

    render() {
        const props = {
            onSubmit: this.handleSubmit,
            errors: this.state.errors
        };

        return React.cloneElement(this.props.children, {...props});
    }

}