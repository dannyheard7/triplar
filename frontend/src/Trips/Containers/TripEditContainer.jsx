import React from "react";
import api from "utils/api.js";

import TripForm from "Trips/Components/TripForm.jsx";
import FormContainer from "../../Forms/Containers/FormContainer";

export default class TripEditContainer extends React.Component {
    constructor(props) {
        super(props);

        this.onSuccess = this.onSuccess.bind(this);
    }

    onSuccess(trip) {
        this.props.onUpdate(trip);
    };

    apiFunc = api.curryAPIFunc(api.patchTrip, this.props.trip.id);

    render() {
        return (
            <div className="card">
                <FormContainer onSuccess={this.onSuccess} apiFunction={this.apiFunc} >
                    <TripForm submitLabel="Save" trip={this.props.trip}/>
                </FormContainer>
            </div>
        );
    }
}

TripEditContainer.defaultProps = {
    trip: {}
};