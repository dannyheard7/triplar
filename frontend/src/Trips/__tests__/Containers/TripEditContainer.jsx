import React from "react";
import {shallow} from "enzyme";

import ShallowRenderer from 'react-test-renderer/shallow';
const renderer = new ShallowRenderer();

import TripEditContainer from "Trips/Containers/TripEditContainer";
import FormContainer from "Forms/Containers/FormContainer";

jest.mock('utils/api.js');
const api = require('utils/api.js');
const faker = require('faker');

describe('<TripEditContainer />', () => {
    const event = {
        preventDefault: () => {},
        target: {
            value: "",
            name: "",
        },
    };

    test('renders correctly', () => {
        const result = renderer.render(<TripEditContainer />);
        expect(result).toMatchSnapshot();
    });

    test('patchTrip called with correct arguments', () => {
        const spy = jest.spyOn(api.default, 'patchTrip');
        const trip = {id: faker.random.number(), name: faker.random.word()};

        const container = shallow(<TripEditContainer  trip={trip} />);
        container.instance().apiFunc(trip);

        expect(spy).toBeCalledWith(trip.id, trip);

        spy.mockRestore();
    });

    test('calls onSuccess after receiving success from <FormContainer />', (function () {
        const stub = jest.spyOn(TripEditContainer.prototype, 'onSuccess');

        const container = shallow(<TripEditContainer onUpdate={jest.fn()}/>);
        container.find(FormContainer).prop('onSuccess')(event);

        expect(stub).toBeCalled();
    }));
});