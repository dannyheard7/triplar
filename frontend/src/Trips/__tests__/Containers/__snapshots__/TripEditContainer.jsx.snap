// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripEditContainer /> renders correctly 1`] = `
<div
  className="card"
>
  <FormContainer
    apiFunction={[Function]}
    onSuccess={[Function]}
  >
    <TripForm
      submitLabel="Save"
      trip={Object {}}
    />
  </FormContainer>
</div>
`;
